`ResponseBuilder::json` requiring the argument to impl `Deref` is confusing (4.0.0-beta.3)

## Expected Behavior

It should be possible to use an owned value for `HttpResponse::Ok().json(<here>)`. Currently, this doesn't work, and the error message is not very helpful:

```
error[E0277]: the trait bound `MySerializableType: Deref` is not satified
```

## Current Behavior

Currently, only types implementing `Deref` can be passed to `Json`.

## Possible Solution

Only require `Serialize` again. `&T where T: Serialize` implements `Serialize` itself so that would work, otherwise requiring a reference like `serde_json::to_string` would work too (it would mean churn when you moved the value into `.json()` before, but at least give a helpful diagnostic). `AsRef<T> where T: Serialize` would work too if you want to be more generic (i.e. supporting smart pointers that don't implement `Serialize`).

## Context

Upgrading from actix-web 3.0 to 4.0.0-beta.3.

## Your Environment

* Rust Version (I.e, output of `rustc -V`): rustc 1.50.0 (cb75ad5db 2021-02-10)
* Actix Web Version: `4.0.0-beta.3`

--- a/actix-http/CHANGES.md
+++ b/actix-http/CHANGES.md
@@ -4,6 +4,7 @@
 ### Changed
 * Feature `cookies` is now optional and disabled by default. [#1981]
 * `ws::hash_key` now returns array. [#2035]
+* `ResponseBuilder::json` now takes `impl Serialize`. [#2052]
 
 ### Removed
 * re-export of `futures_channel::oneshot::Canceled` is removed from `error` mod. [#1994]
@@ -12,6 +13,7 @@
 [#1981]: https://github.com/actix/actix-web/pull/1981
 [#1994]: https://github.com/actix/actix-web/pull/1994
 [#2035]: https://github.com/actix/actix-web/pull/2035
+[#2052]: https://github.com/actix/actix-web/pull/2052
 
 
 ## 3.0.0-beta.3 - 2021-02-10

--- a/actix-http/src/response.rs
+++ b/actix-http/src/response.rs
@@ -5,7 +5,6 @@ use std::{
     convert::TryInto,
     fmt,
     future::Future,
-    ops,
     pin::Pin,
     str,
     task::{Context, Poll},
@@ -673,12 +672,8 @@ impl ResponseBuilder {
     /// Set a json body and generate `Response`
     ///
     /// `ResponseBuilder` can not be used after this call.
-    pub fn json<T>(&mut self, value: T) -> Response
-    where
-        T: ops::Deref,
-        T::Target: Serialize,
-    {
-        match serde_json::to_string(&*value) {
+    pub fn json(&mut self, value: impl Serialize) -> Response {
+        match serde_json::to_string(&value) {
             Ok(body) => {
                 let contains = if let Some(parts) = parts(&mut self.head, &self.err) {
                     parts.headers.contains_key(header::CONTENT_TYPE)
@@ -1007,7 +1002,12 @@ mod tests {
 
     #[test]
     fn test_json() {
-        let resp = Response::build(StatusCode::OK).json(vec!["v1", "v2", "v3"]);
+        let resp = Response::Ok().json(vec!["v1", "v2", "v3"]);
+        let ct = resp.headers().get(CONTENT_TYPE).unwrap();
+        assert_eq!(ct, HeaderValue::from_static("application/json"));
+        assert_eq!(resp.body().get_ref(), b"[\"v1\",\"v2\",\"v3\"]");
+
+        let resp = Response::Ok().json(&["v1", "v2", "v3"]);
         let ct = resp.headers().get(CONTENT_TYPE).unwrap();
         assert_eq!(ct, HeaderValue::from_static("application/json"));
         assert_eq!(resp.body().get_ref(), b"[\"v1\",\"v2\",\"v3\"]");

--- a/src/data.rs
+++ b/src/data.rs
@@ -5,6 +5,7 @@ use std::sync::Arc;
 use actix_http::error::{Error, ErrorInternalServerError};
 use actix_http::Extensions;
 use futures_util::future::{err, ok, LocalBoxFuture, Ready};
+use serde::Serialize;
 
 use crate::dev::Payload;
 use crate::extract::FromRequest;
@@ -99,6 +100,18 @@ impl<T: ?Sized> Clone for Data<T> {
 impl<T: ?Sized> From<Arc<T>> for Data<T> {
     fn from(arc: Arc<T>) -> Self {
         Data(arc)
+    }
+}
+
+impl<T> Serialize for Data<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize(serializer)
     }
 }
 

--- a/src/types/form.rs
+++ b/src/types/form.rs
@@ -106,6 +106,18 @@ impl<T> ops::DerefMut for Form<T> {
     }
 }
 
+impl<T> Serialize for Form<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
 /// See [here](#extractor) for example of usage as an extractor.
 impl<T> FromRequest for Form<T>
 where

--- a/src/types/json.rs
+++ b/src/types/json.rs
@@ -114,6 +114,18 @@ where
     }
 }
 
+impl<T> Serialize for Json<T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
 /// Creates response with OK status code, correct content type header, and serialized JSON payload.
 ///
 /// If serialization failed

